### PR TITLE
feat: [PL-43742]: Send TaskResponse events

### PR DESCRIPTION
**Pull request by mdmilic**
## Describe your changes
Send appropriate event framework events when we get task responses from delegate. 
Send the responses legacy way as well while CD doesn't transition to new response flow. 

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/57121)

### CODE REVIEW

**Comment by mdmilic** [comment]
trigger uts

File: 420-delegate-service/src/main/java/io/harness/taskresponse/TaskResponseService.java [hunk 1]
```diff
@@ -30,15 +40,80 @@
...
+
+    // In case of init failing, acknowledge the failure to delegate and send failure response to task submitter
+    if (response.getResponseCode() != ResponseCode.RESPONSE_OK) {
+      log.error("Error response from delegate {} for execution {}. {}", delegateId, taskId, response.getErrorMessage());
+      final var callbackResponse =
+          InitializeExecutionInfraResponse
+              .builder(task.getUuid(), io.harness.delegate.beans.scheduler.ExecutionStatus.FAILED)
+              .errorMessage(response.getErrorMessage())
+              .build();
+      handleLegacyResponse(task, callbackResponse);
```
Comment (xingchi-jin) [inline 420-delegate-service/src/main/java/io/harness/taskresponse/TaskResponseService.java:59]: can we wrap line 59,60 into a private function ?

**Comment by mdmilic** [comment]
trigger uts

**Comment by mdmilic** [comment]
trigger messagecheck

File: 360-cg-manager/src/main/java/software/wings/app/WingsApplication.java [hunk 2]
```diff
@@ -976,6 +977,7 @@ public GrpcServerConfig getGrpcServerConfig() {
         return configuration.getGrpcServerConfig();
       }
     });
+    modules.add(new TaskResponseEventModule(configuration.getRedisAtmosphereConfig()));
```
Comment (xingchi-jin) [inline 360-cg-manager/src/main/java/software/wings/app/WingsApplication.java:980]: should we create a new config section for "bijouResponseRedisConfig" ? name it in your preference

File: 420-delegate-service/src/main/java/io/harness/taskresponse/event/TaskResponseEventProducer.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    } catch (EventsFrameworkDownException e) {
+      log.error(
+          "Failed to send TaskResponseEvent to events framework for account {} and task {}: ", accountId, taskId, e);
+      return false;
+    }
+    return true;
+  }
+}
+
+//  @Provides
```
Comment (xingchi-jin) [inline 420-delegate-service/src/main/java/io/harness/taskresponse/event/TaskResponseEventProducer.java:66]: remove

File: 420-delegate-service/src/main/java/io/harness/taskresponse/event/TaskResponseCallbackMapper.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+import org.mapstruct.Mapping;
+import org.mapstruct.factory.Mappers;
+
+@Mapper(config = ProtobufMapperConfig.class, uses = StandardProtobufMappers.class)
+public interface TaskResponseCallbackMapper {
+  TaskResponseCallbackMapper INSTANCE = Mappers.getMapper(TaskResponseCallbackMapper.class);
+
+  @Mapping(source = "uuid", target = "taskId.id")
+  @Mapping(source = "code", target = "status")
+  //  @Mapping(target = "sentAt", expression = "java( java.time.Instant.now(); )")
```
Comment (xingchi-jin) [inline 420-delegate-service/src/main/java/io/harness/taskresponse/event/TaskResponseCallbackMapper.java:25]: remove

**Review by xingchi-jin — DISMISSED** [review]
approve with nits

**Comment by mdmilic** [comment]
trigger uts

**Comment by mdmilic** [comment]
trigger uts

**Review by xingchi-jin — APPROVED** [review]

**Pull request merged**